### DISCUSSION TITLE
How to fetch the same data from different components and avoid duplicated requests?

### DISCUSSION THREAD

**Discussion by sumerokr** [nuxt/nuxt#20394] [Questions]
Hi there,
Reproduction demo: [CodeSandbox](https://codesandbox.io/p/github/sumerokr/nuxt3multy-fetch-showcase/master).
Imagine a situation when you want to show the same data in a few different components.
It could be the same menu links for the header and footer.
Or a user's data that is displayed in the main content + a user name in the header and so on.
These components are not aware of each other and cannot rely on the assumption that the data will be fetched by another component.
Therefore to ensure the data is available they initiate a data requests.

### Example
```
// componentA
const { data: users } = await useFetch("https://jsonplaceholder.typicode.com/users");

// componentB
const { data: users } = await useFetch("https://jsonplaceholder.typicode.com/users");
```
If you do that, you see two duplicated requests to the server, both during SSR and client navigation.
Even worse, they are _sequential_ (executed one by one) if "disable cache" is unchecked in DevTools (default userland behavior).
The render (during client navigation) happens in a sequential style as well, they pop up on the screen.

### `key` for the win?
Documentation says that `useFetch` supports (via underlying `useAsyncData`)
> key: a unique key to ensure that data fetching can be properly de-duplicated across requests

Let's try
```
// componentA
const { data: users } = await useFetch("https://jsonplaceholder.typicode.com/users", { key: "users" });

// componentB
const { data: users } = await useFetch("https://jsonplaceholder.typicode.com/users", { key: "users" });
```
If you do that, you _still_ see two duplicated requests to the server, both during SSR and on the client.
They are _still sequential_ (executed one by one) if "disable cache" is unchecked in DevTools (default userland behavior).
The only difference is that they render _all at once_ when the last one is resolved.

----

This feels a bit confusing to me. useAsyncData suppose to have an underlying cache logic that should help with de-duplication.
I wish it could return the same promise that was already created for the same following requests.

What is the best practice to achieve this relatively simple goal?

**Comment by peterroe** [comment]
My thinking is this：
1. You could try to extract the same logic to a common component if you detect the de-duplicate code in your project
2. It is also a good choice to try to explictly store the data locally, suce as `pinia`
3. Maybe you want to use [mem](https://github.com/sindresorhus/mem)

**Comment by sumerokr** [comment]
Hi @peterroe, thanks for the response.

> You could try to extract the same logic to a common component if you detect the de-duplicate code in your project

I assume you meant "duplicated code", right? This could do the trick, but you might have a different representation of this data. E.g. ComponentA renders user names, while ComponentB renders user emails. Sure, we can go further and make it abstract renderless component with scoped slot, but that feels too much already :)

> It is also a good choice to try to explictly store the data locally, suce as pinia

Sure, we can do something like
```
// componentA
const users = usersStore.users;

// componentB
const users = usersStore.users;
```
But someone should kick of and `await` fetching. A Layout, or app.vue? What if these components are not rendered (by some `v-if` condition)? Then we fetch for no reason. Feels a bit wrong to me.

----
It feels that the only way to solve it is to implement a custom de-duplication logic that returns existing promise for the same request.
This is unfortunate, since `useFetch` and `useAsyncData` already have a caching layer that just doesn't help here.

**Comment by peterroe** [comment]
The component encapsulation I'm referring to could be：
```vue
<template>
	<shardComponent v-slot="{ dataA }">
		<yourCompoentA :data="dataA">
	</shardComponent>
</template>
```
```vue
<template>
	<shardComponent v-slot="{ dataB }">
		<yourCompoentB :data="dataB">
	</shardComponent>
</template>
```

**Comment by TheAlexLichter** [comment]
You might want to check https://github.com/nuxt/nuxt/discussions/19912

**Reply by sumerokr** [reply]
@manniL thanks for the response. I'll implement something similar on my side.
Maybe you can give us a hint, what is NuxtData used for internally? I mean we can access it directly, ok. But how can we leverage it via `useFetch` or `useAsyncData` if they always fire a new request?

**Reply by TheAlexLichter** [reply]
@sumerokr you can e.g. write an own composable that checks the current data first or put that in `useAsyncData` even.

See e.g. https://twitter.com/henrycunh/status/1630671452239196160

**Reply by sumerokr** [reply]
@manniL  I hit "mark as answer" too early.
All your examples from above describe how to reuse _already fetched data_.
But my initial question is about a different scenario.
When many components try to fetch the same data at the same time, how to avoid duplicated requests?
E.g. when 10 components initialize request for the same data, and this data is not yet fetched, there is nothing to reuse.
But that doesn't mean that we want to want to actually send 10 network requests.
They are all the same, we want to send just one.
Something like
```TS
const promiseCache: Record<string, any> = {};

export const useDedupeFetch: typeof useAsyncData = (...args) => {
  const [key] = args;

  if (typeof key === 'string' && promiseCache[key]) {
    // cache hit
    return promiseCache[key];
  }

  // cache miss
  const promise = useAsyncData(...args);

  if (typeof key === 'string') {
    promiseCache[key] = promise;

    promise.then(() => {
      // remove as soon as resolved
      delete promiseCache[key];
    });
  }

  return promise;
};

```

**Reply by sumerokr** [reply]
When the code from the prev. message will not work?
Imagine a situation when one of components wants to `watch` for the reactive changes.
```vue
// compA
const langRef = useLang();
const { data: user } = useDedupeFetch(`{langRef.value}-user`, () => $fetch(`/${langRef.value}/some/user/1`));

// compB
const langRef = useLang();
const { data: user } = useDedupeFetch(`{langRef.value}-user`, () => $fetch(`/${langRef.value}/some/user/1`));
  watch: [langRef]
});
```

Now compA is a hostage of compB.
When langRef will be changed (e.g. from `"en"` to `"de"`), compB will re-fetch the data for the new language, because of `watch` option.
But the `key` stays the same. It cannot be dynamic. That means for the above code you could have a German-localized  user for the `"en-user"` cache key. Now think about the fact, that on the server these cache keys are shared between ALL users (as I understand) and feel the complexity of SSR approach :smiley:

**Comment by virgosoy** [comment]
Hello, I hope `effectScope` can help you.
https://github.com/vuejs/rfcs/blob/master/active-rfcs/0041-reactivity-effect-scope.md#example-a-shared-composable

When I used composables like:
```js
// useFetchMyData.js
export async function useFetchMyData(){
  const { data: users } = await useFetch("https://jsonplaceholder.typicode.com/users", {server: false});
  return {
     users
  }
}

// comp1 script setup
const { users } = await useFetchMyData()
// comp2 script setup
const { users } = await useFetchMyData()
// app.vue use comp1 and comp2
```
will has 2 request.

I use effectScope modify useFetchMyData.js
```js
// useFetchMyData.js
async function _useFetchMyData(){
  const { data: users } = await useFetch("https://jsonplaceholder.typicode.com/users", {server: false});
  return {
     users
  }
}
export const useFetchMyData = createSharedComposable(_useFetchMyData)
// createSharedComposable see https://github.com/vuejs/rfcs/blob/master/active-rfcs/0041-reactivity-effect-scope.md#example-a-shared-composable
```
has only one request.
